pyaccel/lifetime.py

```diff
@@ -313,45 +449,179 @@ def touschek_data(self):
+            betagamma2 = (beta*gamma)**2
+
+            val1 = 1/espread2
+            val2 = (eta1*eta1 + eta1til2)/(sig1b2)
+            val3 = (eta2*eta2 + eta2til2)/(sig2b2)
+            sigh2 = 1/(val1 + val2 + val3)
+
+            c1_ = beta1**2/sig1b2*(1-sigh2*eta1til2/sig1b2)
+            c2_ = beta2**2/sig2b2*(1-sigh2*eta2til2/sig2b2)
+            ch_ = (sig1*sig2)**2 - (espread2*eta1*eta2)**2
+
+            b1_ = (c1_ + c2_)/(2*betagamma2)
+            b2_ = (c1_ - c2_)**2/4
+            b2_ += eta1til2*eta2til2*(sigh2/emit1/emit2)**2
+            b2_ /= (betagamma2**2)
+            b2_ = _np.sqrt(b2_)
+
+            taum_p = (beta*d_accp)**2
+            taum_n = (beta*d_accn)**2
+
+            f_int_p = self.f_integral_2_simps(taum_p, b1_, b2_)
+            f_int_n = self.f_integral_2_simps(taum_n, b1_, b2_)
+
+            touschek_coeffs['b1'] = b1_
+            touschek_coeffs['b2'] = b2_
+            touschek_coeffs['taum_p'] = taum_p
+            touschek_coeffs['taum_n'] = taum_n
+            touschek_coeffs['f_int_p'] = f_int_p
+            touschek_coeffs['f_int_n'] = f_int_n
+
+            rate_const = const * nr_part/gamma**2/bunlen
+            rate_const /= _np.sqrt(ch_)
+            ratep = rate_const * f_int_p/taum_p
+            raten = rate_const * f_int_n/taum_n
+            rate = (ratep + raten)/2
+
+        rate = _np.array(rate).ravel()
+        # Average inverse Touschek Lifetime
         avg_rate = _np.trapz(rate, x=s_calc) / (s_calc[-1] - s_calc[0])
-        return dict(rate=rate, avg_rate=avg_rate, volume=vol, pos=s_calc)
+        dit = dict(
+            rate=rate, avg_rate=avg_rate,
+            volume=vol, pos=s_calc,
+            touschek_coeffs=touschek_coeffs)
+        return dit
+
+    @staticmethod
+    def f_function_arg_1(tau, taum, b1_, b2_):
+        """."""
+        ratio = tau/taum/(1+tau)
+        arg = (2+1/tau)**2 * (ratio - 1)
+        arg += 1 - _np.sqrt(1/ratio)
+        arg -= 1/2/tau*(4 + 1/tau) * _np.log(ratio)
+        arg *= _np.sqrt(ratio*taum)
+        bessel = _np.exp(-(b1_-b2_)*tau)*_special.i0e(b2_*tau)
+        return arg * bessel
+
+    @staticmethod
+    def f_integral_1(taum, b1_, b2_):
+        """."""
+        lim = 1000
+        f_int, _ = _integrate.quad(
+            func=Lifetime.f_function_arg_1, a=taum, b=_np.inf,
+            args=(taum, b1_, b2_), limit=lim)
+        f_int *= _np.sqrt(_np.pi*(b1_**2-b2_**2))*taum
+        return f_int
+
+    @staticmethod
+    def f_function_arg_2(kappa, kappam, b1_, b2_):
+        """."""
+        tau = (_np.tan(kappa)**2)[:, None]
+        taum = (_np.tan(kappam)**2)[None, :]
+        ratio = tau/taum/(1+tau)
+        arg = (2*tau+1)**2 * (ratio - 1)/tau
+        arg += tau - _np.sqrt(tau*taum*(1+tau))
+        arg -= (2+1/(2*tau))*_np.log(ratio)
+        arg *= _np.sqrt(1+tau)
+        bessel = _np.exp(-(b1_-b2_)*tau)*_special.i0e(b2_*tau)
+        return arg * bessel
+
+    @staticmethod
+    def f_integral_2(taum, b1_, b2_):
```

Comment:
This guy is not vectorialized is it? I think you have to loop here, since the arguments may be vectors.

Reply:
I decided to remove the use of scipy quadrature and use only the Simpson's rule for the integration. Therefore, this method do not exist anymore.